staging/src/k8s.io/apiserver/pkg/admission/metrics.go

```diff
@@ -0,0 +1,151 @@
+
+	// Built-in admission controller metrics. Each admission controller is identified by name.
+	controller := newMetricSet("controller",
+		[]string{"name", "type", "operation", "group", "version", "resource", "subresource", "rejected"},
+		"Admission controller %s, identified by name and broken out for each operation and API resource and type (validating or mutating).")
+
+	// External admission webhook metrics. Each webhook is identified by name.
+	externalWebhook := newMetricSet("external_webhook",
+		[]string{"name", "type", "operation", "group", "version", "resource", "subresource", "rejected"},
+		"External admission webhook %s, identified by name and broken out for each operation and API resource and type (validating or mutating).")
+
+	step.mustRegister()
+	controller.mustRegister()
+	externalWebhook.mustRegister()
+	return &AdmissionMetrics{step: step, controller: controller, externalWebhook: externalWebhook}
+}
+
+func (m *AdmissionMetrics) reset() {
+	m.step.reset()
+	m.controller.reset()
+	m.externalWebhook.reset()
+}
+
+// ObserveAdmissionStep records admission related metrics for a admission step, identified by step type.
+func (m *AdmissionMetrics) ObserveAdmissionStep(elapsed time.Duration, rejected bool, attr Attributes, stepType string) {
+	gvr := attr.GetResource()
+	m.step.observe(elapsed, stepType, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), strconv.FormatBool(rejected))
+}
+
+// ObserveAdmissionController records admission related metrics for a built-in admission controller, identified by it's plugin handler name.
+func (m *AdmissionMetrics) ObserveAdmissionController(elapsed time.Duration, rejected bool, handler NamedHandler, attr Attributes, stepType string) {
+	gvr := attr.GetResource()
+	m.controller.observe(elapsed, handler.Name(), stepType, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), strconv.FormatBool(rejected))
+}
+
+// ObserveExternalWebhook records admission related metrics for a external admission webhook.
+func (m *AdmissionMetrics) ObserveExternalWebhook(elapsed time.Duration, rejected bool, hook *v1alpha1.Webhook, attr Attributes) {
+	t := "validating" // TODO: pass in type (validating|mutating) once mutating webhook functionality has been implemented
+	gvr := attr.GetResource()
+	m.externalWebhook.observe(elapsed, hook.Name, t, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), strconv.FormatBool(rejected))
+}
+
+type metricSet struct {
+	latencies        *prometheus.HistogramVec
+	latenciesSummary *prometheus.SummaryVec
+}
+
+func newMetricSet(name string, labels []string, helpTemplate string) *metricSet {
+	return &metricSet{
+		latencies: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      fmt.Sprintf("%s_latencies", name),
+				Help:      fmt.Sprintf(helpTemplate, "latency histogram"),
```

Comment:
Same question as for the metric name, latency of what?

Reply:
Latency of *admission*. See answer above.

Reply:
Then please swap the subject and action, i.e. not `admission_controller_latencies`, but `controller_admission_latencies`, because it's not `request_client_latencies` but `client_request_latencies`

Reply:
Maybe you don't even need it in the name, you can make a label out of it

Anyways, next time you have some doubts, please ask someone from the sig-intrumetation to participate in the discussion

/cc @piosz @fgrzadkowski

Reply:
Label is a good idea, I've re-purposed the 'type' label slightly so that it is either _validate_ or _admit_, which should clarify this.

Thanks for the review!

Reply:
Thank you! That really helps!